PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx

```diff
@@ -0,0 +1,381 @@
+		end;
+		
+		[Enum.HumanoidStateType.Running] = function()	
+			local sound = Sounds[SFX.Running]
+			verifyAndSetLoopedForSound(sound,true)
+			stopActiveLoopedSoundsExcept(sound)
+			
+			if Util.HorizontalVelocityMagnitude(Head) > 0.5 then
+				if not sound.IsPlaying then
+					Util.Resume(sound)
+				end
+				setSoundInActiveLooped(sound)
+			else
+				stopActiveLoopedSoundsExcept()
+			end
+		end;
+		
+		[Enum.HumanoidStateType.Swimming] = function()
+			if activeState ~= Enum.HumanoidStateType.Swimming and Util.VerticalVelocityMagnitude(Head) > 0.1 then
+				local splashSound = Sounds[SFX.Splash]
+				splashSound.Volume = Util.Clamp(
+					Util.YForLineGivenXAndTwoPts(
+						Util.VerticalVelocityMagnitude(Head), 
+						100, 0.28, 
+						350, 1),
+					0,1)
+				Util.Play(splashSound)
+			end
+			
+			do
+				local sound = Sounds[SFX.Swimming]
+				verifyAndSetLoopedForSound(sound,true)
+				stopActiveLoopedSoundsExcept(sound)
+				if not sound.IsPlaying then
+					Util.Resume(sound)
+				end
+				setSoundInActiveLooped(sound)
+			end
+		end;
+		
+		[Enum.HumanoidStateType.Climbing] = function()
+			local sound = Sounds[SFX.Climbing]
+			verifyAndSetLoopedForSound(sound,true)
+			if Util.VerticalVelocityMagnitude(Head) > 0.1 then
+				if not sound.IsPlaying then
+					Util.Resume(sound)
+				end				
+				stopActiveLoopedSoundsExcept(sound)
+			else
+				stopActiveLoopedSoundsExcept()
+			end		
+			setSoundInActiveLooped(sound)
+		end;
+		
+		[Enum.HumanoidStateType.Jumping] = function()
+			if activeState == Enum.HumanoidStateType.Jumping then
+				return
+			end		
+			stopActiveLoopedSoundsExcept()
+			local sound = Sounds[SFX.Jumping]
+			verifyAndSetLoopedForSound(sound,false)
+			Util.Play(sound)
+		end;
+		
+		[Enum.HumanoidStateType.GettingUp] = function()
+			stopActiveLoopedSoundsExcept()
+			local sound = Sounds[SFX.GettingUp]
+			verifyAndSetLoopedForSound(sound,false)
+			Util.Play(sound)
+		end;
+		
+		[Enum.HumanoidStateType.Freefall] = function()
+			if activeState == Enum.HumanoidStateType.Freefall then
+				return
+			end
+			local sound = Sounds[SFX.FreeFalling]
+			if sound.Volume ~= 0 then
+				sound.Volume = 0
+			end
+			stopActiveLoopedSoundsExcept()
+		end;
+		
+		[Enum.HumanoidStateType.FallingDown] = function()
+			stopActiveLoopedSoundsExcept()
+		end;
+		
+		[Enum.HumanoidStateType.Landed] = function()
+			stopActiveLoopedSoundsExcept()
+			if Util.VerticalVelocityMagnitude(Head) > 75 then
+				local landingSound = Sounds[SFX.Landing]
+				landingSound.Volume = Util.Clamp(
+					Util.YForLineGivenXAndTwoPts(
+						Util.VerticalVelocityMagnitude(Head), 
+						50, 0, 
+						100, 1),
+					0,1)
+				Util.Play(landingSound)			
+			end
+		end
+	}
+	
+	-- Handle state event fired or OnChange fired
+	function stateUpdated(state)
+		if stateUpdateHandler[state] ~= nil then
+			stateUpdateHandler[state]()
+		end
+		activeState = state
+	end
+	
+	-- Runs on heartbeat
+	function onHeartbeat(step)
```

Comment:
can we not do this with events? it seems like it would be more performant to not need a loop here. If a loop is necessary why not spawn or delay a function in the freefall state that does this?